contributors/design-proposals/scheduling/scheduling-framework.md

```diff
@@ -0,0 +1,406 @@
+These plugins have one more chance to check any conditions about the assignment
+of the Pod to this Node and reject the node if needed.
+
+![image](images/scheduling-framework-threads.png)
+
+### Reserve
+
+This is not a plugin point. At this point scheduler updates its cache by
+"reserving" a Node (partially or fully) for the Pod. In scheduler v1 this stage
+is called "assume". At this point, only the scheduler cache is updated to
+reflect that the Node is (partially) reserved for the Pod. The actual assignment
+of the Node to the Pod happens during the "Bind" phase. That is when the API
+server updates the Pod object with the Node information.
+
+### Admit
+
+Admit plugins run in a separate go routine (in parallel). Each plugin can return
+one of the three possible values: 1) "admit", 2) "reject", or 3) "wait". If all
+plugins registered at this extension point return "admit", the pod is sent to
+the next step for binding. If any of the plugins returns "reject", the pod is
+rejected and sent back to the scheduling queue. If any of the plugins returns
+"wait", the Pod is kept in reserved state until it is explicitly approved for
+binding. A plugin that returns "wait" must return a "timeout" as well. If the
+timeout expires, the pod is rejected and goes back to the scheduling queue.
+
+#### Approving a Pod binding
+
+While any plugin can receive the list of reserved Pod from the cache and approve
+them,  we expect only the "Admit" plugins to approve binding of reserved Pods
+that are in "waiting" state. Once a Pod is approved, it is sent to the Bind
+stage.
+
+### Reject
+
+Plugins called at "Admit" may perform some operations that should be undone if
+the Pod reservation fails. The "Reject" extension point allows such clean-up
+operations to happen. Plugins registered at this point are called if the
+reservation of the Pod is cancelled. The reservation is cancelled if any of the
+"Admit" plugins returns "reject" or if a Pod reservation, which is in "wait"
+state, times out. 
+
+### Pre-Bind
+
+When a Pod is approved for binding it reaches to this stage. These plugins run
+before the actual binding of the Pod to a Node happens. The binding starts only
+if all of these plugins return true. If any returns false, the Pod is rejected
+and sent back to the scheduling queue. These plugins run in a separate go
+routine. The same go routine runs "Bind" after these plugins when all of them
+return true.
+
+### Bind
+
+Once all pre-bind plugins return true, the Bind plugins are executed. Multiple
+plugins may be registered at this extension point. Each plugin may return true
+or false (or an error). If a plugin returns false, the next plugin will be
+called until a plugin returns true. Once a true is returned **the remaining
+plugins are skipped**. If any of the plugins returns an error or all of them
+return false, the Pod is rejected and sent back to the scheduling queue.
+
+### Post Bind
```

Comment:
Curious if there is any use-case for this extension point

Reply:
If bind fails, certain things may need to be undone. For example, if we dynamically bind volumes to a node at "pre-bind" and then bind to that node fails, we probably want to detach the volumes.